File: examples/plotting/file/text.py
from bokeh.layouts import column
from bokeh.models import CustomJS, Slider
from bokeh.plotting import figure, show, output_file

aligns    = ["left", "center", "right"]
baselines = ["bottom", "middle", "top" ]

p = figure(x_range=aligns, y_range=baselines, plot_width=800, plot_height=600,
           title="Variations of multi-line text", toolbar_location=None)

p.background_fill_color = "lightgrey"
p.xaxis.axis_label = "align"
p.yaxis.axis_label = "baseline"
p.axis.major_label_text_font_size = "18px"
p.axis.major_label_text_font_style = "bold"

p.xgrid.grid_line_color = None
p.ygrid.grid_line_color = None

texts = [

'''one''',

'''two
lines''',

'''lines
here:
3''',

'''here
are
4
lines'''

]

def xs(cat):
    return [(cat, -0.3), (cat, -0.1), (cat, 0.1), (cat, 0.3)]

def ys(cat):
    return [cat] * 4

renderers = {}
i = 0

for a in aligns:
    for b in baselines:
        r = p.text(xs(a), ys(b), texts, text_align=a, text_baseline=b,
                   text_font_size="14px", text_line_height=1.2)
        renderers["r" + str(i)] = r
        i += 1

slider = Slider(title="Text Angle", start=0, end=45, step=1, value=0)
slider.js_on_change('value', CustomJS(args=renderers, code="""
    rs = [r0, r1, r2 , r3, r4, r5, r6, r7, r8];
    for (i=0; i<9; i++) {
        rs[i].glyph.angle = {value: cb_obj.value, units: "deg"}
    }
"""))

output_file("text.html")
show(column(p, slider))

File: bokehjs/test/integration/glyphs/text.ts
import {display, fig} from "../_util"
import type {TextAlign, TextBaseline} from "@bokehjs/core/enums"

describe("Text glyph", () => {
  it("should allow to paint text with embedded scripts", async () => {
    const p = fig([250, 100], {x_range: [0, 16], y_range: [0, 2]})
    p.text(0, 0, ["</script><script>throw new Error('XSS1')</script>"])
    p.text(0, 1, ["<script>throw new Error('XSS0')</script>"])
    await display(p)
  })

  describe("should support multi-line rendering with different alignments", () => {
    function plot(angle: number) {
      const aligns    = ["left", "center", "right"] as const
      const baselines = ["bottom", "middle", "top"] as const

      const p = fig([800, 400], {x_range: aligns, y_range: baselines})

      p.xaxis.axis_label = "align"
      p.yaxis.axis_label = "baseline"
      p.yaxis.major_label_orientation = "parallel"
      p.axis.major_label_text_font_size = "12px"
      p.axis.major_label_text_font_style = "bold italic"

      p.xgrid.grid_line_color = null
      p.ygrid.grid_line_color = null

      const texts = [
        "one",
        "two\nlines",
        "lines\nhere:\n3",
        "here\nare\n4\nlines",
      ]

      const x4 = <T>(cat: T) => [cat, cat, cat, cat]

      function xs(cat: string): [string, number][] {
        return [[cat, -0.3], [cat, -0.1], [cat, 0.1], [cat, 0.3]]
      }

      const x = []
      const y = []
      const text = []
      const text_align: TextAlign[] = []
      const text_baseline: TextBaseline[] = []
      const text_color = []

      for (const align of aligns) {
        for (const baseline of baselines) {
          x.push(...xs(align))
          y.push(...x4(baseline))
          text.push(...texts)
          text_align.push(...x4(align))
          text_baseline.push(...x4(baseline))
          text_color.push("black", "gray", "blue", "green")
        }
      }

      p.text({
        x: x as any, // TODO: offset factors
        y,
        text,
        text_align,
        text_baseline,
        text_color,
        angle: {value: angle, units: "deg"},
        text_font_size: "10px",
      })

      return p
    }

    for (const angle of [-30, 0, 30, 60]) {
      it(`at ${angle} degree angle`, async () => {
        const p = plot(angle)
        await display(p)
      })
    }
  })
})
